authLdap.php

```diff
@@ -362,13 +360,17 @@ function authLdap_login($user, $username, $password, $already_md5 = false)
                 $user_info['user_email'] = $username . '@example.com';
             }
             $userid = wp_insert_user($user_info);
+            if (is_wp_error($userid)) {
+                authLdap_debug('Error creating user : ' . $userid->get_error_message());
+                trigger_error('Error creating user: ' . $userid->get_error_message());
+                return $userid;
+            }
         }
 
-        // if the user exists, wp_insert_user will update the existing user record
-        if (is_wp_error($userid)) {
-            authLdap_debug('Error creating user : ' . $userid->get_error_message());
-            trigger_error('Error creating user: ' . $userid->get_error_message());
-            return $userid;
+        $user = new WP_User($userid);
+        $user->set_role(''); // removes all existing roles.
+        foreach ($roles as $role) {
+            $user->add_role($role);
```

Comment:
Yes, I think that could be done more gracefully.  Would need to add new roles and remove old ones.  I figured it wasn't much different than making a call to wp_update_user().